docs/developer/best-practices/security.asciidoc

```diff
@@ -1,55 +1,128 @@
-Pollution by checking `Object.hasOwnProperty` if the arguments to the
-APIs originate from an Object. An example is the Code app’s
-https://github.com/elastic/kibana/blob/b49192626a8528af5d888545fb14cd1ce66a72e7/x-pack/legacy/plugins/code/server/lsp/workspace_command.ts#L40-L44[spawnProcess].
-*** Common Node.js offenders: `child_process.spawn`,
-`child_process.exec`, `eval`, `Function('some string')`,
-`vm.runIn*Context(x)`
-*** Common Client-side offenders: `eval`, `Function('some string')`,
-`setTimeout('some string', num)`, `setInterval('some string', num)`
-* Check for accidental reveal of sensitive information
-** The biggest culprit is errors which contain stack traces or other
-sensitive information which end up in the HTTP Response
-* Checked for Mishandled API requests
-** Ensure no sensitive cookies are forwarded to external resources.
-** Ensure that all user controllable variables that are used in
-constructing a URL are escaped properly. This is relevant when using
-`transport.request` with the {es} client as no automatic
-escaping is performed.
-* Reverse tabnabbing -
-https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/HTML5_Security_Cheat_Sheet.md#tabnabbing
-** When there are user controllable links or hard-coded links to
-third-party domains that specify target="_blank" or target="_window", the a tag should have the rel="noreferrer noopener" attribute specified.
-Allowing users to input markdown is a common culprit, a custom link renderer should be used
-* SSRF - https://www.owasp.org/index.php/Server_Side_Request_Forgery
-All network requests made from the {kib} server should use an explicit configuration or white-list specified in the kibana.yml
\ No newline at end of file
+When writing code for {kib}, be sure to follow these best practices to avoid common vulnerabilities. Refer to the included Open Web
+Application Security Project (OWASP) references to learn more about these types of attacks.
+
+=== Cross-site Scripting (XSS) ===
+
+https://owasp.org/www-community/attacks/xss[_OWASP reference for XSS_]
+
+XSS is a class of attacks where malicious scripts are injected into vulnerable websites. {kib} defends against this primarily by widespread
+use of the React framework to safely encode data that is rendered to pages, widespread use of the EUI framework to
+https://elastic.github.io/eui/#/navigation/link#link-validation[automatically sanitize links], and a restrictive Content-Security-Policy
+header.
+
+*Best practices:*
+
+* Check for usage of dangerous functions or assignments that can result in unescaped user input in the browser DOM. Avoid
+using these at all costs:
+** *React:* https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml[`dangerouslySetInnerHtml`].
+** *Browser DOM:* `Element.innerHTML` and `Element.outerHTML`.
+** *jQuery:* `$.html`, `$.append`, `$.appendTo`, `$.prepend`, and `$.prependTo`. Consider using `$.text` instead, which is a safe
+alternative to these functions.
```

Comment:
I debated removing this bit about jQuery since it seems that we're not using it anywhere anymore. Thoughts?

Reply:
I'm fine with removing it.